### PR TITLE
Euler import connector

### PR DESCRIPTION
**Pull request by shriyatyagii**
Added Euler import connector

image

### CODE REVIEW

File: contracts/mainnet/connectors/euler/euler-import/main.sol [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+
+		_eventName = "LogEulerImport(address,uint,uint,address[],address[],bool[])";
+		_eventParam = abi.encode(
+			userAccount,
+			sourceId,
+			targetId,
+			inputData._supplyTokens,
+            data.supplyAmts,
+			inputData._borrowTokens,
+            data.borrowAmts,
```
Comment (thrilok209) [inline contracts/mainnet/connectors/euler/euler-import/main.sol]:
```suggestion
			data.supplyAmts,
			inputData._borrowTokens,
			data.borrowAmts,
```

File: contracts/mainnet/connectors/euler/euler-import/main.sol [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+			k++;
+		}
+
+		address[] memory deferLiquidityChecks = new address[](2);
+		deferLiquidityChecks[0] = helper.sourceAccount;
+		deferLiquidityChecks[1] = helper.targetAccount;
+
+		eulerExec.batchDispatch(items, deferLiquidityChecks);
+
+		_eventName = "LogEulerImport(address,uint,uint,address[],address[],bool[])";
```
Comment (thrilok209) [inline contracts/mainnet/connectors/euler/euler-import/main.sol]:
```suggestion
		_eventName = "LogEulerImport(address,uint256,uint256,address[],uint256[],address[],uint256[],bool[])";
```

File: contracts/mainnet/connectors/euler/euler-import/main.sol [hunk 3]
```diff
@@ -0,0 +1,157 @@
+//SPDX-License-Identifier: MIT
+pragma solidity ^0.7.0;
+pragma experimental ABIEncoderV2;
+import "./helpers.sol";
+import "./interface.sol";
+import "./events.sol";
+
+contract EulerImport is EulerHelpers {
```
Comment (thrilok209) [inline contracts/mainnet/connectors/euler/euler-import/main.sol]: @shriyatyagii Can you once lint the code? Few lines are out of indentation

File: contracts/mainnet/connectors/euler/euler-import/main.sol [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+
+        // BorrowAmts will be in underlying token decimals
+		data = getBorrowAmounts(helper.sourceAccount, inputData, data);
+
+		// SupplyAmts will be in 18 decimals
+		data = getSupplyAmounts(helper.sourceAccount, inputData, data);
+
+		helper.supplylength = data.supplyTokens.length;
+		helper.borrowlength = data.borrowTokens.length;
+		uint16 enterMarkets = 0;
```
Comment (thrilok209) [inline contracts/mainnet/connectors/euler/euler-import/main.sol]:
```suggestion
		uint16 enterMarketsLength = 0;
```

File: contracts/mainnet/connectors/euler/euler-import/main.sol [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+	function _importEuler(
+		address userAccount,
+		uint256 sourceId,
+		uint256 targetId,
+		ImportInputData memory inputData
+	) 
+        internal 
+        returns (string memory _eventName, bytes memory _eventParam) 
+    {
+
```
Comment (pradyuman-verma) [inline contracts/mainnet/connectors/euler/euler-import/main.sol]: We also need check that `sourceId` and `targetId` should be less than 256.

File: contracts/mainnet/connectors/euler/euler-import/helpers.sol [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+			}
+		}
+		for (uint256 i = 0; i < inputData._supplyTokens.length; i++) {
+			address _token = inputData._supplyTokens[i] == ethAddr
+				? wethAddr
+				: inputData._supplyTokens[i];
+			data.supplyTokens[i] = _token;
+			data.eTokens[i] = EulerTokenInterface(markets.underlyingToEToken(_token));
+			data.supplyAmts[i] = data.eTokens[i].balanceOf(userAccount);//All 18 dec
+		}
```
Comment (pradyuman-verma) [inline contracts/mainnet/connectors/euler/euler-import/helpers.sol]:
```suggestion
                uint256 length_ = inputData._supplyTokens.length;
		for (uint256 i = 0; i < length_; i++) {
			for (uint256 j = i + 1; j < length_; j++) {
					require(
						inputData._supplyTokens[i] != inputData._supplyTokens[j],
						"token-repeated"
					);
			}
		}
		for (uint256 i = 0; i < length_; i++) {
			address _token = inputData._supplyTokens[i] == ethAddr
				? wethAddr
				: inputData._supplyTokens[i];
			data.supplyTokens[i] = _token;
			data.eTokens[i] = EulerTokenInterface(markets.underlyingToEToken(_token));
			data.supplyAmts[i] = data.eTokens[i].balanceOf(userAccount);//All 18 dec
		}
```

File: contracts/mainnet/connectors/euler/euler-import/helpers.sol [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+			data.borrowAmts = new uint256[](_borrowTokensLength);
+			for (uint256 i = 0; i < _borrowTokensLength; i++) {
+				for (uint256 j = i; j < _borrowTokensLength; j++) {
+					if (j != i) {
+						require(
+							inputData._borrowTokens[i] !=
+								inputData._borrowTokens[j],
+							"token-repeated"
+						);
+					}
```
Comment (pradyuman-verma) [inline contracts/mainnet/connectors/euler/euler-import/helpers.sol]:
```suggestion
				for (uint256 j = i + 1; j < _borrowTokensLength; j++) {
						require(
							inputData._borrowTokens[i] !=
								inputData._borrowTokens[j],
							"token-repeated"
						);
```

File: contracts/mainnet/connectors/euler/euler-import/main.sol (hunk 3 above)
Comment (shriyatyagii) [reply contracts/mainnet/connectors/euler/euler-import/main.sol]: I did lint the code, not sure what happened. I'll do it once again

File: contracts/mainnet/connectors/euler/euler-import/main.sol (hunk 5 above)
Comment (shriyatyagii) [reply contracts/mainnet/connectors/euler/euler-import/main.sol]: Added

File: contracts/mainnet/connectors/euler/euler-import/helpers.sol [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+		}
+		for (uint256 i = 0; i < length_; i++) {
+			address _token = inputData._supplyTokens[i] == ethAddr
+				? wethAddr
+				: inputData._supplyTokens[i];
+			data.supplyTokens[i] = _token;
+			data.eTokens[i] = EulerTokenInterface(
+				markets.underlyingToEToken(_token)
+			);
+			data.supplyAmts[i] = data.eTokens[i].balanceOf(userAccount); //All 18 dec
```
Comment (pradyuman-verma) [inline contracts/mainnet/connectors/euler/euler-import/helpers.sol]: @shriyatyagii It should be `balanceOfUnderlying` or `balanceOf`?

File: contracts/mainnet/connectors/euler/euler-import/helpers.sol [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+	}
+
+	function getBorrowAmounts(
+		address userAccount, // user's EOA sub-account address
+		ImportInputData memory inputData,
+		ImportData memory data
+	) internal view returns (ImportData memory) {
+		uint256 _borrowTokensLength = inputData._borrowTokens.length;
+
+		if (_borrowTokensLength > 0) {
```
Comment (pradyuman-verma) [inline contracts/mainnet/connectors/euler/euler-import/helpers.sol]: We use this standard for memory variables: 

```
Constant/Immutables storage variables: VAR_NAME.
Upgradable storage variables: _varName. (that’s how OpenZepplin does as well for storage variables)
Memory variables: varName_. (any memory variable: whether it’s a param, return or a variable defined inside function)
```
```suggestion
		uint256 borrowTokensLength_ = inputData._borrowTokens.length;

		if (borrowTokensLength_ > 0) {
```

File: contracts/mainnet/connectors/euler/euler-import/helpers.sol [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+			data.borrowTokens = new address[](_borrowTokensLength);
+			data.dTokens = new EulerTokenInterface[](_borrowTokensLength);
+			data.borrowAmts = new uint256[](_borrowTokensLength);
+			for (uint256 i = 0; i < _borrowTokensLength; i++) {
+				for (uint256 j = i + 1; j < _borrowTokensLength; j++) {
+					require(
+						inputData._borrowTokens[i] !=
+							inputData._borrowTokens[j],
+						"token-repeated"
```
Comment (pradyuman-verma) [inline contracts/mainnet/connectors/euler/euler-import/helpers.sol]:
```suggestion
						"borrow-token-repeated"
```

File: contracts/mainnet/connectors/euler/euler-import/helpers.sol [hunk 11]
```diff
@@ -0,0 +1,127 @@
...
+		data.supplyTokens = new address[](inputData._supplyTokens.length);
+		data.eTokens = new EulerTokenInterface[](
+			inputData._supplyTokens.length
+		);
+		uint256 length_ = inputData._supplyTokens.length;
+		for (uint256 i = 0; i < length_; i++) {
+			for (uint256 j = i + 1; j < length_; j++) {
+				require(
+					inputData._supplyTokens[i] != inputData._supplyTokens[j],
+					"token-repeated"
```
Comment (pradyuman-verma) [inline contracts/mainnet/connectors/euler/euler-import/helpers.sol]:
```suggestion
					"supply-token-repeated"
```

File: contracts/mainnet/connectors/euler/euler-import/main.sol [hunk 12]
```diff
@@ -0,0 +1,158 @@
...
+				allowError: false,
+				proxyAddr: address(data.eTokens[i]),
+				data: abi.encodeWithSignature(
+					"transferFrom(address,address,uint256)",
+					helper.sourceAccount,
+					helper.targetAccount,
+					data.supplyAmts[i]
+				)
+			});
+			k++;
```
Comment (pradyuman-verma) [inline contracts/mainnet/connectors/euler/euler-import/main.sol]:
```suggestion
			items[k++] = IEulerExecute.EulerBatchItem({
				allowError: false,
				proxyAddr: address(data.eTokens[i]),
				data: abi.encodeWithSignature(
					"transferFrom(address,address,uint256)",
					helper.sourceAccount,
					helper.targetAccount,
					data.supplyAmts[i]
				)
			});
```

File: contracts/mainnet/connectors/euler/euler-import/main.sol [hunk 13]
```diff
@@ -0,0 +1,158 @@
...
+				items[k] = IEulerExecute.EulerBatchItem({
+					allowError: false,
+					proxyAddr: address(markets),
+					data: abi.encodeWithSignature(
+						"enterMarket(uint256,address)",
+						targetId,
+						data.supplyTokens[i]
+					)
+				});
+				k++;
```
Comment (pradyuman-verma) [inline contracts/mainnet/connectors/euler/euler-import/main.sol]:
```suggestion
				items[k++] = IEulerExecute.EulerBatchItem({
					allowError: false,
					proxyAddr: address(markets),
					data: abi.encodeWithSignature(
						"enterMarket(uint256,address)",
						targetId,
						data.supplyTokens[i]
					)
				});
```

File: contracts/mainnet/connectors/euler/euler-import/main.sol [hunk 14]
```diff
@@ -0,0 +1,158 @@
...
+				allowError: false,
+				proxyAddr: address(data.dTokens[j]),
+				data: abi.encodeWithSignature(
+					"transferFrom(address,address,uint256)",
+					helper.sourceAccount,
+					helper.targetAccount,
+					data.borrowAmts[j]
+				)
+			});
+			k++;
```
Comment (pradyuman-verma) [inline contracts/mainnet/connectors/euler/euler-import/main.sol]:
```suggestion
			items[k++] = IEulerExecute.EulerBatchItem({
				allowError: false,
				proxyAddr: address(data.dTokens[j]),
				data: abi.encodeWithSignature(
					"transferFrom(address,address,uint256)",
					helper.sourceAccount,
					helper.targetAccount,
					data.borrowAmts[j]
				)
			});
```

File: contracts/mainnet/connectors/euler/euler-import/helpers.sol (hunk 11 above)
Comment (thrilok209) [reply contracts/mainnet/connectors/euler/euler-import/helpers.sol]: @shriyatyagii We can remove this logic.

File: contracts/mainnet/connectors/euler/euler-import/helpers.sol (hunk 10 above)
Comment (thrilok209) [reply contracts/mainnet/connectors/euler/euler-import/helpers.sol]: @shriyatyagii Same here. We can remove these. As we used it for INST LM safety checks

File: contracts/mainnet/connectors/euler/euler-import/helpers.sol (hunk 9 above)
Comment (shriyatyagii) [reply contracts/mainnet/connectors/euler/euler-import/helpers.sol]: Will keep this thing in mind!

File: contracts/mainnet/connectors/euler/euler-import/helpers.sol (hunk 10 above)
Comment (shriyatyagii) [reply contracts/mainnet/connectors/euler/euler-import/helpers.sol]: removed

File: contracts/mainnet/connectors/euler/euler-import/helpers.sol (hunk 11 above)
Comment (shriyatyagii) [reply contracts/mainnet/connectors/euler/euler-import/helpers.sol]: removed

File: contracts/mainnet/connectors/euler/euler-import/helpers.sol (hunk 8 above)
Comment (shriyatyagii) [reply contracts/mainnet/connectors/euler/euler-import/helpers.sol]: @pradyuman-verma We can technically use any. But `balanceOf` will return eToken value in 18 decimals. If we use `balanceOfUnderlying`, we'll have to convert everything to 18 decimals and then pass this 18 decimals amount in `transferFrom`.

File: contracts/mainnet/connectors/euler/euler-import/helpers.sol (hunk 11 above)
Comment (shriyatyagii) [reply contracts/mainnet/connectors/euler/euler-import/helpers.sol]: Removed

File: contracts/mainnet/connectors/euler/euler-import/main.sol [hunk 15]
```diff
@@ -0,0 +1,155 @@
...
+			data.supplyAmts,
+			inputData._borrowTokens,
+			data.borrowAmts,
+			inputData._enterMarket
+		);
+	}
+}
+
+contract ConnectV2EulerImport is EulerImport {
+	string public constant name = "Euler-import-v1.0";
```
Comment (thrilok209) [inline contracts/mainnet/connectors/euler/euler-import/main.sol]:
```suggestion
	string public constant name = "Euler-Import-v1.0";
```

**Comment by shriyatyagii** [comment]
Connector Details-

Name: `EULER-IMPORT-A`
Address: [0x3860AB3E6dBB46a267cbEe72D5Fb81ca07Ff6daf](https://etherscan.io/address/0x3860AB3E6dBB46a267cbEe72D5Fb81ca07Ff6daf#code)
Transaction - [tx](https://etherscan.io/tx/0xa3f3efc0d92808afa844289e7b28511ebe2b5432b963ba430c23fb0a70d85b42)

**Review by thrilok209 — APPROVED** [review]

**Pull request merged**